src/System.Buffers.Primitives/System/Buffers/OutputWriter.cs

```diff
@@ -35,79 +35,42 @@ public void Advance(int count)
             _output.Advance(count);
         }
 
-        public void Write(byte[] source)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Write(ReadOnlySpan<byte> source)
         {
-            if (source.Length > 0 && _span.Length >= source.Length)
+            if (_span.Length >= source.Length)
             {
-                ref byte pSource = ref source[0];
-                ref byte pDest = ref MemoryMarshal.GetReference(_span);
-
-                Unsafe.CopyBlockUnaligned(ref pDest, ref pSource, (uint)source.Length);
-
+                source.CopyTo(_span);
                 Advance(source.Length);
             }
             else
             {
-                WriteMultiBuffer(source, 0, source.Length);
+                WriteMultiBuffer(source);
             }
         }
 
-        public void Write(byte[] source, int offset, int length)
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void Ensure(int count = 1)
         {
-            // If offset or length is negative the cast to uint will make them larger than int.MaxValue
-            // so each test both tests for negative values and greater than values. This pattern wil also
-            // elide the second bounds check that would occur at source[offset]; as is pre-checked
-            // https://github.com/dotnet/coreclr/pull/9773
-            if ((uint)offset > (uint)source.Length || (uint)length > (uint)(source.Length - offset))
-            {
-                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.offset);
-            }
-
-            if (length > 0 && _span.Length >= length)
-            {
-                ref byte pSource = ref source[offset];
-                ref byte pDest = ref MemoryMarshal.GetReference(_span);
-
-                Unsafe.CopyBlockUnaligned(ref pDest, ref pSource, (uint)length);
-
-                Advance(length);
-            }
-            else
-            {
-                WriteMultiBuffer(source, offset, length);
-            }
+            _output.Enlarge(count);
```

Comment:
Not related to this PR, but we need to make sure we standardize Enlarge/Ensure (naming) and what the parameter means.